### PR TITLE
feat: Added multiple tiles support in `get_edge_population`

### LINKED ISSUE #124: bug: `add_edge_population` fails is there are two GHSL tiles overlapping the study area
Hi @cbueth 

When the study area overlaps more than one of the GHSL tiles, the function `add_edge_population` returns an error.

Here is a reprex for this behaviour

```python
import osmnx as ox
from superblockify.population.approximation import add_edge_population

boundary_lat, boundary_lon = 41.8956456, 12.4655397

graph = ox.graph_from_point(
    (boundary_lat, boundary_lon), 
    dist=500, 
    network_type='drive', 
    simplify=True
)

graph = ox.add_edge_speeds(graph)
graph = ox.add_edge_travel_times(graph)
graph = ox.project_graph(graph)

add_edge_population(graph, overwrite=True)
# 2026-04-01 12:21:54,478 |     INFO | tessellation.py:101 | Calculating edge cells for graph with 165 edges.
# 2026-04-01 12:21:54,609 |     INFO | tessellation.py:155 | Tessellated 132 edge cells in 0:00:00.
# 2026-04-01 12:21:54,672 |     INFO | ghsl.py:129 | Using the GHSL raster tiles for the bounding box (1039130.3128053441, 4999534.908906185, 1040472.7243228266, 5000784.670403941).
# TypeError: invalid path or file: ['/home/juan/create-toy-nets/data/ghsl/GHS_POP_E2025_GLOBE_R2023A_54009_100_V1_0_R4_C20.tif', '/home/juan/create-toy-nets/data/ghsl/GHS_POP_E2025_GLOBE_R2023A_54009_100_V1_0_R5_C20.tif']
# Cell In[56], line 1
# ----> 1 add_edge_population(graph, overwrite=True)

```

To fix this, we will need to preprocess the tiles beforehand. The changes shouldn't be too long

### PR DESCRIPTION
**Pull request by juanfonsecaLS1**
## Description

Adds a function to handle multiple GHSL tile files when the study area spans across more than one tile.

Small change to the code in `get_edge_population`

## Related Issue

Fixes #124

## Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 📝 Documentation update
- [ ] 🧹 Code refactor (no functional changes)
- [ ] ✅ Test update

## How Has This Been Tested?

- [ ] Unit tests
- [ ] Integration tests
- [x] Manual testing

I used the same minimal example in Rome
```
import osmnx as ox
from superblockify.population.approximation import add_edge_population
import geopandas as gpd

boundary_lat, boundary_lon = 41.8956456, 12.4655397

graph = ox.graph_from_point(
    (boundary_lat, boundary_lon), 
    dist=500, 
    network_type='drive', 
    simplify=True
)

graph = ox.project_graph(graph)

add_edge_population(graph, overwrite=True)

```

## Screenshots (if applicable)

image

## Checklist

- [ ] My code follows the project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix/feature works
- [x] New and existing tests pass locally

### CODE REVIEW

**Comment by cbueth** [comment]
Thanks for the commits @juanfonsecaLS1 and working on this improvement.
I think I would prefer it to have this list check inside `load_ghsl_as_polygons`, not adding a separate function. Maybe `load_ghsl_as_polygons` can call this stitching rasters together inside and then calling itself. Do you get the idea? Is there any reason making it unpractical?
Would you please add a test that tests this literal edge case?
Furthermore, there has been a commit on main that you may want to merge.
Greetings

**Comment by juanfonsecaLS1** [comment]
Hi @cbueth, I initially went for the vrt approach as I mentioned before. However, that implied either having an extra dependency in the package or building a XML to create the mosaic. 
In the end I went for your initial suggestion, contenating the polygons returned by `load_ghsl_as_polygons`. I preferred not to change the `window` argument that `load_ghsl_as_polygons`. 

```python
        with rasopen(ghsl_file) as src:
            load_window = src.window(*bbox_moll)
        
        ghsl_polygons = load_ghsl_as_polygons(ghsl_file, window=load_window)
```

I did not manage to find a better way round this part of the code. I imagine we can just incorporate the code directly into the `get_edge_population` if we want to avoid the new function. What do you think?

 Sure, I can add a test with the Rome case.

Also, I will rebase this branch too.

Thanks!

File: superblockify/population/approximation.py [hunk 1]
```diff
@@ -6,6 +6,8 @@
 from functools import partial
 from multiprocessing import Pool
 
+import pandas as pd
+import geopandas as gpd
```
Comment (cbueth) [inline superblockify/population/approximation.py]: Would you import the function needed separately instad of loading the whole packages? You can see below how with the import from `geopandas`.

File: superblockify/population/approximation.py [hunk 2]
```diff
@@ -124,6 +126,50 @@ def get_population_area(graph):
...
+    """
+    ghsl_polygons_list = []
+
+    for ghsl_file in ghsl_files:
+        logger.debug("Loading GHSL polygons from tile: %s", ghsl_file)
+        with rasopen(ghsl_file) as src:
+            load_window = src.window(*bbox_moll)
+        # Load polygons from the windowed region of this tile
+        ghsl_polygons = load_ghsl_as_polygons(ghsl_file, window=load_window)
+        ghsl_polygons_list.append(ghsl_polygons)
```
Comment (cbueth) [inline superblockify/population/approximation.py:155]: When loading the polygons and concatenating. What happens with the polygons at the border? If a street polygon goes through both rasters. Does the population get summed at any point?

File: superblockify/population/approximation.py [hunk 3]
```diff
@@ -185,10 +231,13 @@ def get_edge_population(graph, batch_size=10000, **tess_kwargs):
     bbox_moll = edge_cells.union_all().buffer(100).bounds
     ghsl_file = get_ghsl(bbox_moll)
 
-    with rasopen(ghsl_file) as src:
-        load_window = src.window(*bbox_moll)
-
-    ghsl_polygons = load_ghsl_as_polygons(ghsl_file, window=load_window)
+    # Checking if ghsl_file is a list of multiple tile files. See Issue #124.
```
Comment (cbueth) [inline superblockify/population/approximation.py]: No need to cite the issue number here.

**Review by cbueth — CHANGES_REQUESTED** [review]
Understood, that is a fine choice I think.

Yes, please add two test cases that fall on two polygons. You can parametrize it with `@pytest.mark.parametize()`.

The black code style is failing, this is some formatting needed to be applied, if possible.

The failing tests in the current CI are due to timeouts I see. I think we need to find an alternative so the tests do not all need to download the tile directly from the server as they are too many resquests, it seems. But this is a separate issue.

File: superblockify/population/approximation.py (hunk 3 above)
Comment (juanfonsecaLS1) [reply superblockify/population/approximation.py]: Updated

File: superblockify/population/approximation.py (hunk 1 above)
Comment (juanfonsecaLS1) [reply superblockify/population/approximation.py]: Yes, I missed this. I will fix this

File: superblockify/population/approximation.py (hunk 2 above)
Comment (juanfonsecaLS1) [reply superblockify/population/approximation.py:155]: I don't think this part of the code deals with that. My guess is that it would be equivalent to have a road segment/edge spanning over multiple raster cells. I agree that it's worth checking the results for those links to ensure consistent results.

Comment (cbueth) [reply superblockify/population/approximation.py:155]: You are right, I mixed up this step with the one distibuting the population/weighted summing.

Comment (juanfonsecaLS1) [reply superblockify/population/approximation.py:155]: Yes, just double checked and the `_population_fraction_list` already handles such cases.

File: superblockify/population/approximation.py (hunk 1 above)
Comment (juanfonsecaLS1) [reply superblockify/population/approximation.py]: updated

Comment (cbueth) [reply superblockify/population/approximation.py]: thanks!

**Comment by juanfonsecaLS1** [comment]
Just added a parametrised test with a sample of two locations near the edge of a tile. Just checked the results of the Black test and it seems it is an access issue:
```
remote: Permission to BikeNetKit/superblockify.git denied to github-actions[bot].
fatal: unable to access 'https://github.com/BikeNetKit/superblockify/': The requested URL returned error: 403
```

**Review by cbueth — CHANGES_REQUESTED** [review]
Sorry for the late reply.

Thanks for the changes. It is nearly ready to merge. For now we can ignore the black CI step. This I need to fix separately.

I would suggest to add explicit tile verification to `test_get_edge_population_multiple_tiles`.

The test doesn't verify that Rome and Turbo actually span multiple GHSL tiles. Add this before the add_edge_population call:
```python
from geopandas import GeoDataFrame
from superblockify.population.ghsl import get_ghsl
# ... inside test function after project_graph
boundary_gdf = GeoDataFrame(
    geometry=[graph.graph["boundary"]], crs=graph.graph["boundary_crs"]
).to_crs("World Mollweide")
ghsl_result = get_ghsl(bbox_moll=list(boundary_gdf.total_bounds))
assert isinstance(ghsl_result, list), (
    f"{city_name} should span multiple GHSL tiles, but get_ghsl returned "
    f"{type(ghsl_result).__name__} instead of list"
)
assert len(ghsl_result) >= 2, (
    f"{city_name} should need at least 2 GHSL tiles, but get_ghsl returned "
    f"{len(ghsl_result)} tile(s)"
)
```
Why: I do believe you that these two examples span multiple tiles, but without this, if the multi-tile handling breaks (e.g., `_load_ghsl_multifile` is removed or `get_ghsl` changes to always return a single path), the test would still pass but wouldn't actually test the multi-tile case.
